scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,407 @@
+            streams_table_name: "system_distributed.cdc_streams_descriptions_v2".to_string(),
+            session: Arc::clone(session),
+        }
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_all_stream_generations_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps';
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_all_generations(&self) -> Result<Vec<Generation>, Box<dyn Error>> {
+        let mut rows = self
+            .session
+            .query_iter(self.get_all_stream_generations_query(), &[])
+            .await?
+            .into_typed::<(chrono::Duration,)>();
+
+        let mut generations = vec![];
+
+        while let Some(next_row) = rows.next().await {
+            let (generation,) = next_row?;
+            generations.push(Generation(generation));
+        }
+
+        Ok(generations)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_generation_by_timeuuid_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time <= dateOf( ? )
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_generation_by_timeuuid(
+        &self,
+        time: &Uuid,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let result = self
+            .session
+            .query(self.get_generation_by_timeuuid_query(), (time,))
+            .await?
+            .rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_next_generation_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps'
+    AND time > ?
+    ORDER BY time DESC
+    LIMIT 1;
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_next_generation(
+        &self,
+        generation: &Generation,
+    ) -> Result<Option<Generation>, Box<dyn Error>> {
+        let result = self
+            .session
+            .query(self.get_next_generation_query(), (Timestamp(generation.0),))
+            .await?
+            .rows;
+
+        GenerationFetcher::return_single_row(result)
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_stream_ids_by_time_query(&self) -> String {
+        format!(
+            r#"
+    SELECT streams
+    FROM {}
+    WHERE time = ?;
+    "#,
+            self.streams_table_name
+        )
+    }
+
+    // Internal function configurable for testing purposes.
+    pub async fn fetch_stream_ids(
```

Comment:
This should work as well xD

```
pub async fn fetch_stream_ids(
        &self,
        timestamp: &Generation,
    ) -> Result<Vec<StreamID>, Box<dyn Error>> {
        Ok(self.session.query_iter(
            self.get_stream_ids_by_time_query(),
            (Timestamp(timestamp.0), ),
        ).await?.map(|x| StreamID(x.unwrap().into_typed::<(Vec<u8>, )>().unwrap().0)).collect().await)
    }

Reply:
I don't think so since each row contains `set<blob>` and cql set's equivalent is `Vec<T>` and cql blob's equivalent is `Vec<u8>`, so it ought to be `[...].into_typed::<(Vec<Vec<u8>>, )>()[...]`. I had some problem when trying to implement it, but it could be because of my lack of understanding of Rust.

Reply:
> each row contains `set<blob>`
Would you be so kind to point me to the specific part of documentation supporting this claim?

Reply:
The table we are querying is `system_distributed.cdc_streams_descriptions_v2`. Using cqlsh we can check its schema using `DESCRIBE`. It gives following output:
```
cqlsh> describe system_distributed.cdc_streams_descriptions_v2;

CREATE TABLE system_distributed.cdc_streams_descriptions_v2 (
    time timestamp,
    range_end bigint,
    streams frozen<set<blob>>,
    PRIMARY KEY (time, range_end)
) WITH CLUSTERING ORDER BY (range_end ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.0
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';
```

Reply:
Yeah, you're right, thank you.

Just for fun: this one works but it's admittedly messier ;)
```rust
pub async fn fetch_stream_ids(
        &self,
        timestamp: &Generation,
    ) -> Result<Vec<StreamID>, Box<dyn Error>> {
        Ok(self.session.query_iter(
            self.get_stream_ids_by_time_query(),
            (Timestamp(timestamp.0), ),
        ).await?.into_typed::<(Vec<Vec<u8>>, )>().flat_map(
            |x| futures::stream::iter(x.unwrap().0.iter().map(|y| StreamID(y.clone())).collect::<Vec<_>>())
        ).collect().await)
    }
```